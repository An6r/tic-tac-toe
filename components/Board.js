import React from 'react';
import { PropTypes } from 'prop-types';
import BoardCell from './BoardCell';

const Board = ({ grid, handleCellClick, isWinnerMark, getPlayerSymbol }) => (
  <ul className='grid'>
    {grid.map((rows, rowColumn) =>
      rows.map((row, rowIndex) => (
        <BoardCell key={`row-${rowIndex}-${rowColumn}`}
                   column={rowColumn}
				   value={row}
				   index={rowIndex}
				   handleCellClick={handleCellClick}
				   isWinnerMark={isWinnerMark}
				   getPlayerSymbol={getPlayerSymbol}
        />
      )),
    )}
  </ul>
);

Board.propTypes = {
  grid: PropTypes.arrayOf(
    PropTypes.arrayOf(
      PropTypes.number,
    ).isRequired,
  ).isRequired,
  handleCellClick: PropTypes.func.isRequired,
  /**
   * A function to check if the board space is in the winner marks.
   */
  isWinnerMark: PropTypes.func.isRequired,
  getPlayerSymbol: PropTypes.func.isRequired
};

export default Board;