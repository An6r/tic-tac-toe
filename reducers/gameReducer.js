import _ from 'lodash';
import * as actionTypes from '../actions/actionTypes';
import {
  BOARD_SIZE,
  generateBoardGrid,
  verifyBoardCells
} from '../helpers/game';

export const initialState = {
  boardGrid: generateBoardGrid(),
  currentPlayer: 1,
  playersSymbols: {
	  1: 'X',
	  2: 'O'
  },
  winnerMarks: {},
  winnerPlayer: null,
  marksCount: 0,
  isTie: false,
};

export const gameReducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.NEW_SIGN: {
      const { currentPlayer } = state;
      const boardGrid = _.cloneDeep(state.boardGrid); 

      const { column, row } = action;

      boardGrid[column][row] = currentPlayer;

      return {
        ...state,
        boardGrid,
        marksCount: state.marksCount + 1,
      };
    }
    case actionTypes.VERIFY_BOARD: {
      const boardSize = BOARD_SIZE;
      const maxMarksCount = boardSize ** 2;

      const {
		  currentPlayer: winnerPlayer,
		  marksCount
	  } = state;

      if (marksCount === maxMarksCount) {
        return {
          ...state,
          isTie: true,
        };
      }

      const boardGrid = _.cloneDeep(state.boardGrid);

      for (let i = 0; i < boardSize; i += 1) {
        // Check if user won horizontally
        if (verifyBoardCells(boardGrid[i][0], boardGrid[i][1], boardGrid[i][2])) {
          return {
            ...state,
            winnerMarks: {
              [i]: [0, 1, 2],
            },
            winnerPlayer,
          };
        }

        // Check if user won vertically
        if (verifyBoardCells(boardGrid[0][i], boardGrid[1][i], boardGrid[2][i])) {
          return {
            ...state,
            winnerMarks: {
              0: [i],
              1: [i],
              2: [i],
            },
            winnerPlayer,
          };
        }
      }

      /*
       Check if user won diagonally
       */
      if (verifyBoardCells(boardGrid[0][0], boardGrid[1][1], boardGrid[2][2])) {
        return {
          ...state,
          winnerMarks: {
            0: [0],
            1: [1],
            2: [2],
          },
          winnerPlayer,
        };
      }

      if (verifyBoardCells(boardGrid[0][2], boardGrid[1][1], boardGrid[2][0])) {
        return {
          ...state,
          winnerMarks: {
            0: [2],
            1: [1],
            2: [0],
          },
          winnerPlayer,
        };
      }

      return state;
    }
    case actionTypes.NEXT_PLAYER: {
      return {
        ...state,
        currentPlayer: ((state.currentPlayer === 1) ? 2 : 1),
      };
    }
    case actionTypes.RESTART_GAME: {
      return {
        ...initialState
      };
    }
    default: {
      return state;
    }
  }
};