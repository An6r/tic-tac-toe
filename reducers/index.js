import { gameReducer as game } from './gameReducer';

export default {
	game
};
