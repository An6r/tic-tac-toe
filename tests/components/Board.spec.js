import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { initialState } from '../../reducers/gameReducer';

import Board from '../../components/Board';

function setupComponent() {
  const props = {
    grid: initialState.boardGrid,
    handleCellClick: jest.fn(),
    isWinnerMark: jest.fn(),
    getPlayerSymbol: jest.fn(),
  };

  const wrapper = shallow(
    <Board {...props} />,
  );

  return {
    props,
    wrapper,
  };
}

describe('<Board />', () => {
  it('should render', () => {
    const { wrapper } = setupComponent();

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
