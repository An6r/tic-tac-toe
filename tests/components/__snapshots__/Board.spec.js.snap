// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> should render 1`] = `
<ul
  className="grid"
>
  <BoardSpace
    column={0}
    getPlayerSymbol={[MockFunction]}
    handleCellClick={[MockFunction]}
    index={0}
    isWinnerMark={[MockFunction]}
    value={null}
  />
  <BoardSpace
    column={0}
    getPlayerSymbol={[MockFunction]}
    handleCellClick={[MockFunction]}
    index={1}
    isWinnerMark={[MockFunction]}
    value={null}
  />
  <BoardSpace
    column={0}
    getPlayerSymbol={[MockFunction]}
    handleCellClick={[MockFunction]}
    index={2}
    isWinnerMark={[MockFunction]}
    value={null}
  />
  <BoardSpace
    column={1}
    getPlayerSymbol={[MockFunction]}
    handleCellClick={[MockFunction]}
    index={0}
    isWinnerMark={[MockFunction]}
    value={null}
  />
  <BoardSpace
    column={1}
    getPlayerSymbol={[MockFunction]}
    handleCellClick={[MockFunction]}
    index={1}
    isWinnerMark={[MockFunction]}
    value={null}
  />
  <BoardSpace
    column={1}
    getPlayerSymbol={[MockFunction]}
    handleCellClick={[MockFunction]}
    index={2}
    isWinnerMark={[MockFunction]}
    value={null}
  />
  <BoardSpace
    column={2}
    getPlayerSymbol={[MockFunction]}
    handleCellClick={[MockFunction]}
    index={0}
    isWinnerMark={[MockFunction]}
    value={null}
  />
  <BoardSpace
    column={2}
    getPlayerSymbol={[MockFunction]}
    handleCellClick={[MockFunction]}
    index={1}
    isWinnerMark={[MockFunction]}
    value={null}
  />
  <BoardSpace
    column={2}
    getPlayerSymbol={[MockFunction]}
    handleCellClick={[MockFunction]}
    index={2}
    isWinnerMark={[MockFunction]}
    value={null}
  />
</ul>
`;
